File: android/app/src/main/java/app/seamlessclip/protocol/SessionCipher.kt
package app.seamlessclip.protocol

import java.nio.ByteBuffer
import java.security.GeneralSecurityException
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Per-connection AES-256-GCM with counter nonces; session key from HKDF over the pairing key
 * and both handshake nonces. Byte-for-byte compatible with SessionCipher.cs.
 *
 * Not thread-safe: callers serialise [seal] (write lock) and [open] (single reader).
 */
class SessionCipher(
    pairingKey: ByteArray,
    serverNonce: ByteArray,
    clientNonce: ByteArray,
    isServer: Boolean,
) {
    private val key: SecretKeySpec
    private val sendDirection: Int
    private val receiveDirection: Int
    private var sendCounter = 0L
    private var receiveCounter = 0L

    init {
        val derived = Hkdf.sha256(pairingKey, serverNonce + clientNonce, HKDF_INFO.toByteArray(Charsets.US_ASCII), 32)
        key = SecretKeySpec(derived, "AES")
        derived.fill(0)
        sendDirection = if (isServer) SERVER_TO_CLIENT else CLIENT_TO_SERVER
        receiveDirection = if (isServer) CLIENT_TO_SERVER else SERVER_TO_CLIENT
    }

    fun seal(plaintext: ByteArray): ByteArray {
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.ENCRYPT_MODE, key, GCMParameterSpec(TAG_BITS, nonce(sendDirection, sendCounter++)))
        return cipher.doFinal(plaintext) // ciphertext || 16-byte tag
    }

    /** @throws GeneralSecurityException (usually AEADBadTagException) if the frame doesn't authenticate. */
    fun open(frame: ByteArray): ByteArray {
        if (frame.size < TAG_BITS / 8) throw GeneralSecurityException("Encrypted frame shorter than GCM tag")
        val cipher = Cipher.getInstance(TRANSFORMATION)
        cipher.init(Cipher.DECRYPT_MODE, key, GCMParameterSpec(TAG_BITS, nonce(receiveDirection, receiveCounter)))
        val plaintext = cipher.doFinal(frame)
        receiveCounter++
        return plaintext
    }

    private fun nonce(direction: Int, counter: Long): ByteArray =
        ByteBuffer.allocate(12).putInt(direction).putLong(counter).array() // big-endian

    companion object {
        const val HKDF_INFO = "seamless-clip v1 session"
        private const val TRANSFORMATION = "AES/GCM/NoPadding"
        private const val TAG_BITS = 128
        private const val SERVER_TO_CLIENT = 1
        private const val CLIENT_TO_SERVER = 2
    }
}

File: windows/SeamlessClip/Protocol/SessionCipher.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace SeamlessClip.Protocol;

/// <summary>
/// Per-connection AES-256-GCM with counter nonces. The session key is derived with
/// HKDF-SHA256 from the pairing key and both handshake nonces. See docs/PROTOCOL.md.
/// Not thread-safe: callers serialize <see cref="Seal"/> (write lock) and <see cref="Open"/> (single reader).
/// </summary>
public sealed class SessionCipher : IDisposable
{
    public const string HkdfInfo = "seamless-clip v1 session";
    public const int TagLength = 16;
    private const int NonceLength = 12;
    private const uint ServerToClient = 1;
    private const uint ClientToServer = 2;

    private readonly AesGcm _aes;
    private readonly uint _sendDirection;
    private readonly uint _receiveDirection;
    private ulong _sendCounter;
    private ulong _receiveCounter;

    public SessionCipher(byte[] pairingKey, byte[] serverNonce, byte[] clientNonce, bool isServer)
    {
        byte[] salt = [.. serverNonce, .. clientNonce];
        byte[] key = HKDF.DeriveKey(HashAlgorithmName.SHA256, pairingKey, 32, salt, Encoding.ASCII.GetBytes(HkdfInfo));
        _aes = new AesGcm(key, TagLength);
        CryptographicOperations.ZeroMemory(key);

        _sendDirection = isServer ? ServerToClient : ClientToServer;
        _receiveDirection = isServer ? ClientToServer : ServerToClient;
    }

    public byte[] Seal(ReadOnlySpan<byte> plaintext)
    {
        Span<byte> nonce = stackalloc byte[NonceLength];
        BuildNonce(nonce, _sendDirection, _sendCounter++);

        var output = new byte[plaintext.Length + TagLength];
        _aes.Encrypt(nonce, plaintext, output.AsSpan(0, plaintext.Length), output.AsSpan(plaintext.Length));
        return output;
    }

    /// <exception cref="CryptographicException">The frame was not produced with this session key/counter.</exception>
    public byte[] Open(ReadOnlySpan<byte> frame)
    {
        if (frame.Length < TagLength)
            throw new CryptographicException("Encrypted frame is shorter than the GCM tag");

        Span<byte> nonce = stackalloc byte[NonceLength];
        BuildNonce(nonce, _receiveDirection, _receiveCounter);

        var plaintext = new byte[frame.Length - TagLength];
        _aes.Decrypt(nonce, frame[..^TagLength], frame[^TagLength..], plaintext);
        _receiveCounter++;
        return plaintext;
    }

    private static void BuildNonce(Span<byte> nonce, uint direction, ulong counter)
    {
        BinaryPrimitives.WriteUInt32BigEndian(nonce, direction);
        BinaryPrimitives.WriteUInt64BigEndian(nonce[4..], counter);
    }

    public void Dispose() => _aes.Dispose();
}
